Source/WebKit/WebProcess/WebCoreSupport/WebPermissionController.cpp

```diff
@@ -86,11 +89,23 @@ void WebPermissionController::removeObserver(WebCore::PermissionObserver& observ
     m_observers.remove(observer);
 }
 
-void WebPermissionController::permissionChanged(const WebCore::ClientOrigin& origin, const WebCore::PermissionDescriptor& descriptor, WebCore::PermissionState state)
+void WebPermissionController::permissionChanged(WebCore::PermissionName permissionName, const WebCore::SecurityOriginData& topOrigin)
 {
     for (auto& observer : m_observers) {
-        if (observer.origin() == origin && observer.descriptor() == descriptor)
-            observer.stateChanged(state);
+        if (observer.origin().topOrigin == topOrigin && observer.descriptor().name == permissionName) {
+            auto& context = observer.context();
+            auto source = WebCore::Permissions::sourceFromContext(context);
+            // FIXME: Add support for workers.
+            ASSERT(source && *source == WebCore::PermissionQuerySource::Window);
+            auto* page = downcast<WebCore::Document>(context).page();
+
+            if (page) {
+                query(WebCore::ClientOrigin { observer.origin() }, WebCore::PermissionDescriptor { permissionName }, page, *source, [observer = WeakPtr { observer }](auto newState) {
+                    if (newState && *newState != observer->currentState())
```

Comment:
You need to check that observer is non-null here since it is a WeakPtr.

Also, the newState check is useless here. The operator==() for std::optional already does this for you. This should be:
```
if (observer && newState != observer->currentState())
```